crates/primitives-evm-tracing-events/src/runtime.rs

```diff
@@ -0,0 +1,343 @@
+//! EVM runtime events definitions.
+
+extern crate alloc;
+
+use codec::{Decode, Encode};
+#[cfg(feature = "evm-tracing")]
+use evm::Opcode;
+pub use evm::{ExitError, ExitReason, ExitSucceed};
+use sp_core::{sp_std::vec::Vec, H160, H256, U256};
+
+use crate::Context;
+#[cfg(feature = "evm-tracing")]
+use crate::StepEventFilter;
+
+/// EVM stack.
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq)]
+pub struct Stack {
+    /// Data.
+    pub data: Vec<H256>,
+    /// Limit.
+    pub limit: u64,
+}
+
+impl From<&evm::Stack> for Stack {
+    fn from(stack: &evm::Stack) -> Self {
+        Self {
+            data: stack.data().clone(),
+            limit: stack.limit() as u64,
+        }
+    }
+}
+
+/// EVM memory.
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq)]
+pub struct Memory {
+    /// Data.
+    pub data: Vec<u8>,
+    /// Effective length.
+    pub effective_len: U256,
+    /// Limit.
+    pub limit: u64,
+}
+
+impl From<&evm::Memory> for Memory {
+    fn from(memory: &evm::Memory) -> Self {
+        Self {
+            data: memory.data().clone(),
+            effective_len: memory.effective_len(),
+            limit: memory.limit() as u64,
+        }
+    }
+}
+
+/// Capture represents the result of execution.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Encode, Decode)]
+pub enum Capture<E, T> {
+    /// The machine has exited. It cannot be executed again.
+    Exit(E),
+    /// The machine has trapped. It is waiting for external information, and can
+    /// be executed again.
+    Trap(T),
+}
+
+/// A type alias representing trap data. Should hold the marshalled `Opcode`.
+pub type Trap = Vec<u8>;
+
+/// EVM runtime event.
+#[derive(Debug, Clone, Encode, Decode, PartialEq, Eq)]
+pub enum RuntimeEvent {
+    /// Step.
+    Step {
+        /// Context.
+        context: Context,
+        /// Opcode.
+        opcode: Vec<u8>,
```

Comment:
Why is `opcode` `Vec<u8>`? Who decides what format goes here? What is this used for?

Reply:
I've just remembered that I tried to use something simpler instead of `Vec<u8>` but faced some issues during next part implementations that are depend on `Vec<u8>` like this one - https://github.com/moonbeam-foundation/moonbeam/blob/02a7c84efffebe47267434e7ddc415f1e11276d6/client/evm-tracing/src/types/mod.rs#L79. 

Then I decided to keep it as is and walk through opcode type refactoring again after having full working version with e2e tests.

What do you think about adding an issue to recheck opcode type usage when we merge full logic to feature branch before master to avoid getting bogged down?

Reply:
- added an issue https://github.com/humanode-network/humanode/issues/1600
- added a task at #1560